### PR TITLE
Manifest V3 support

### PR DESCRIPTION
**Pull request by kdzwinel**
**Reviewer:** @andrey-p 

## Description:
Chrome is introducing significant changes to how extensions work (we refer to those changes as "Manifest V3" or MV3). With many of those changes already implemented and available for testing in the test versions of Chrome we are able to start migrating our existing extension.

## Steps to test this PR:
Full blocking, https upgrading, ATB and whitelisting functionality needs to be tested in both MV3 version of the extension (because it's new) and the MV2/FF version (because this PR modifies some shared code).

Testing MV3 version:
1. Download latest Chromium from https://download-chromium.appspot.com/
2. Build rules and download surrogates:

```
cd scripts/mv3
node fetchTDS.js
node fetchSurrogates.js
node tdsToMV3Rules.js
```

3. Build MV3 extension `npm run dev-chrome-mv3`
4. Load as an unpacked extension

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/classes/tab.es6.js [hunk 1]
```diff
@@ -38,14 +38,14 @@ class Tab {
...
-        this.upgradedHttps = false
+        this.upgradedHttps = tabData.upgradedHttps === true
         this.hasHttpsError = false
         this.mainFrameUpgraded = false
         this.requestId = tabData.requestId
         this.status = tabData.status
         this.site = new Site(this.url)
         this.httpsRedirects = new HttpsRedirects()
-        this.statusCode = null // statusCode is set when headers are recieved in tabManager.js
+        this.statusCode = tabData.statusCode || null // statusCode is set when headers are recieved in tabManager.js
```
Comment (kdzwinel) [inline shared/js/background/classes/tab.es6.js]: since we are serializing and de-serializing TabManager from storage we need a way to set those properties.

File: shared/js/background/load.es6.js [hunk 2]
```diff
@@ -14,32 +14,14 @@ function url (url) {
...
-
 /*
  * Params:
  *  - url: request URL
  *  - source: requests are internal by default. set source to 'external' for non-extension URLs
  *  - etag: set an if-none-match header
  */
 function loadExtensionFile (params) {
-    let xhr = new XMLHttpRequest()
+    const headers = new Headers()
```
Comment (kdzwinel) [inline shared/js/background/load.es6.js]: XMLHttpRequest is not available in service worker, which forces us (thank goodness) to use `fetch`

File: shared/js/background/redirect.es6.js [hunk 3]
```diff
@@ -139,13 +139,11 @@ function handleRequest (requestData) {
                     }
                 }
 
-                if (!window.safari) {
-                    // Initiate hiding of blocked ad DOM elements
-                    tryElementHide(requestData, thisTab)
-                }
+                // Initiate hiding of blocked ad DOM elements
```
Comment (kdzwinel) [inline shared/js/background/redirect.es6.js]: Unrelated change (sorry), we don't support safari anymore so I removed it

File: shared/js/background/tab-manager.es6.js [hunk 4]
```diff
@@ -71,12 +79,17 @@ class TabManager {
...
+
+        if (!tab) {
             info.id = id
             tabManager.create(info)
         } else {
-            let tab = tabManager.get({tabId: id})
-            if (tab && info.status) {
+            if (info.url) {
+                tab.url = info.url
+            }
```
Comment (kdzwinel) [inline shared/js/background/tab-manager.es6.js]: saving to storage at the end and URL upgrading are only changes here, everything else is just a clean up

File: shared/js/background/tab-manager.es6.js [hunk 5]
```diff
@@ -121,6 +137,99 @@ class TabManager {
                 tab.updateSite(request.url)
             }
         }
+
+        this.saveStateToStorage()
+    }
+
+    // if the background page is not persistent (service worker) we need to try to recover
+    // TabManager state from storage
```
Comment (kdzwinel) [inline shared/js/background/tab-manager.es6.js]: We are serializing and deserializing because service worker can be killed by the browser when inactive and we don't want to lose tab state - grade, trackers loaded, etc

File: shared/js/background/utils.es6.js [hunk 6]
```diff
@@ -114,6 +114,10 @@ function getAsyncBlockingSupport () {
     return false
 }
 
+function getPersistentBackgroundPageSupport () {
+    return (typeof chrome.runtime.getBackgroundPage) === 'function'
```
Comment (kdzwinel) [inline shared/js/background/utils.es6.js]: In MV3 there is no access to background page through getBackgroundPage, we need to send messages.

File: shared/js/shared-utils/parse-user-agent-string.es6.js [hunk 7]
```diff
@@ -1,8 +1,13 @@
...
-    if (!uaString) uaString = window.navigator.userAgent
+    if (!uaString) uaString = navigator.userAgent
 
     let browser
     let version
+    let os = 'o'
+
+    if (uaString.indexOf('Windows') !== -1) os = 'w'
+    if (uaString.indexOf('Mac') !== -1) os = 'm'
+    if (uaString.indexOf('Linux') !== -1) os = 'l'
```
Comment (kdzwinel) [inline shared/js/shared-utils/parse-user-agent-string.es6.js]: In MV2 OS is saved in localStorage (only thing we are saving in localStorage 🤷‍♂️). Since localStorage is not available in service worker, instead of storing OS I just extended this method to return info about OS based on UA.

File: shared/js/ui/base/chrome-ui-wrapper.es6.js [hunk 8]
```diff
@@ -28,34 +30,34 @@ let getBackgroundTabData = () => {
     })
 }
 
-let search = (url) => {
-    window.chrome.tabs.create({url: `https://duckduckgo.com/?q=${url}&bext=${window.localStorage['os']}cr`})
+const search = (url) => {
+    const os = parseUserAgentString().os
```
Comment (kdzwinel) [inline shared/js/ui/base/chrome-ui-wrapper.es6.js]: instead of pulling OS from localStorage I'm calling the method that pulls that info from UA

File: shared/js/background/background.es6.js [hunk 9]
```diff
@@ -19,16 +19,4 @@
 const events = require('./$BROWSER-events.es6')
 const settings = require('./settings.es6')
 
-settings.ready().then(() => {
```
Comment (kdzwinel) [inline shared/js/background/background.es6.js]: This is MV2 background page entry point - it's not used in MV3, but since I dropped need for localStorage I cleaned it up.

File: shared/js/background/chrome-mv3-redirect.es6.js [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            frameId = requestData.frameId
+            messageType = 'blockedFrameAsset'
+        }
+
+        chrome.tabs.sendMessage(requestData.tabId, {type: messageType, request: requestData, mainFrameUrl: tab.url}, {frameId: frameId})
+    } else if (!tab.elementHidingDisabled) {
+        chrome.tabs.sendMessage(requestData.tabId, {type: 'disable'})
+        tab.elementHidingDisabled = true
+    }
+}
```
Comment (andrey-p) [inline shared/js/background/chrome-mv3-redirect.es6.js]: Worth deduping this and `isSameDomainRequest`?

File: shared/js/background/chrome-mv3-redirect.es6.js [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+    // get information about the tab
+    const tabManagerTab = tabManager.get(requestData)
+    const isMainFrame = requestData.type === 'main_frame'
+    let isUpgradable = https.canUpgradeUrl(requestData.url)
+
+    // if returned result is an boolean, and not a promise, put it into a resolved promise for easier handling
+    if (!(isUpgradable instanceof Promise)) {
+        isUpgradable = Promise.resolve(isUpgradable)
+    }
```
Comment (andrey-p) [inline shared/js/background/chrome-mv3-redirect.es6.js]: Feels like we should get `canUpgradeUrl` to always return the same value but ah well. This works for now.

File: shared/js/background/allow-pages.es6.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * We have to make sure that domains safelisted by the user, and domains safelisted by our temporary safelist (broken pages)
+ * are in sync with actual 'allow page' rules
+ */
+function syncSafelistEntries () {
+    let res, rej
+    const promise = new Promise((resolve, reject) => { res = resolve; rej = reject })
+
+    chrome.declarativeNetRequest.getAllowedPages(allowedPagesRules => {
```
Comment (andrey-p) [inline shared/js/background/allow-pages.es6.js]: :red_circle: Has the API changed for this? On Chromium 82 I see:

Screenshot from 2020-02-24 16-52-53
